# Lista 5, 6, 7

## Zadanie 2

zad 2 tablice >>> przeprowadzic kilka testow :)))
na podstawie zadania zidentyfikowac jaka jest hipoteza zerowa i jaka jest hipoteza alternatywna????

BINOM.TEST 
PROP.TEST W 2 WERSJACH Z KOREKCJA I BEZ KOREKCJI

To co trzeba zrobic......
zident. h0 ha
zapisac jakie testy :))
fragment kodu z wywolaniem
wynik >> pvalue w kazdym tescie i odpowiednie wnioski na podst pval


---

Niech $\text{Apap}$ oznacza 

In [3]:
data_vec <- c(35, 0, 0, 22, 22, 0, 15, 15, 15, 0, 40, 10, 18, 3, 5)

data <- matrix(data_vec, nrow=5, byrow=T)
data

[,1] [,2] [,3]
[1,] 35    0    0  
[2,] 22   22    0  
[3,] 15   15   15  
[4,]  0   40   10  
[5,] 18    3    5

In [15]:
### DO USUNIECIA
?prop.test
#?binom.test

### a.

Niech $p$ oznacza prawdopodobieństwo, że losowo wybrana osoba z badanej populacji w przypadku bólu zażywa Apap.

W podpunkcie a, hipotezą zerową jest $$H_0: p \leq \frac{1}{4}.$$ Zatem hipotezą alternatywną jest $$H_a: p > \frac{1}{4}.$$

In [10]:
prop.test(44, 200, p = 1/4, alternative = "g")


	1-sample proportions test with continuity correction

data:  44 out of 200, null probability 1/4
X-squared = 0.80667, df = 1, p-value = 0.8154
alternative hypothesis: true p is greater than 0.25
95 percent confidence interval:
 0.1734603 1.0000000
sample estimates:
   p 
0.22 


In [16]:
prop.test(44, 200, p = 1/4, alternative = "g", correct = FALSE)


	1-sample proportions test without continuity correction

data:  44 out of 200, null probability 1/4
X-squared = 0.96, df = 1, p-value = 0.8364
alternative hypothesis: true p is greater than 0.25
95 percent confidence interval:
 0.1757337 1.0000000
sample estimates:
   p 
0.22 


In [14]:
binom.test(44, 200, p = 1/4, alternative = "g")


	Exact binomial test

data:  44 and 200
number of successes = 44, number of trials = 200, p-value = 0.8562
alternative hypothesis: true probability of success is greater than 0.25
95 percent confidence interval:
 0.172679 1.000000
sample estimates:
probability of success 
                  0.22 


W teście `prop.test` bez korekcji p-wartość kształtuje się na poziomie $0.8154 > 0.05$, natomiast z korekcją $0.8364 > 0.05$. W `binom.test` wynosi ona $0.8562 > 0.05$, więc nie ma przesłanek do odrzucenia $H_0$.

### b.

Ponownie, niech $p$ oznacza prawdopodobieństwo, że losowo wybrana osoba z badanej populacji w przypadku bólu zażywa Apap.

W podpunkcie b, hipotezą zerową jest $$H_0: p = \frac{1}{2}.$$ Zatem hipotezą alternatywną jest $$H_a: p \neq \frac{1}{2}.$$

In [11]:
prop.test(44, 200, p = 1/2, alternative = "t")


	1-sample proportions test with continuity correction

data:  44 out of 200, null probability 1/2
X-squared = 61.605, df = 1, p-value = 4.198e-15
alternative hypothesis: true p is not equal to 0.5
95 percent confidence interval:
 0.1659406 0.2850661
sample estimates:
   p 
0.22 


In [17]:
prop.test(44, 200, p = 1/2, alternative = "t", correct = FALSE)


	1-sample proportions test without continuity correction

data:  44 out of 200, null probability 1/2
X-squared = 62.72, df = 1, p-value = 2.383e-15
alternative hypothesis: true p is not equal to 0.5
95 percent confidence interval:
 0.1681654 0.2823880
sample estimates:
   p 
0.22 


In [13]:
binom.test(44, 200, p = 1/2, alternative = "t")


	Exact binomial test

data:  44 and 200
number of successes = 44, number of trials = 200, p-value = 6.838e-16
alternative hypothesis: true probability of success is not equal to 0.5
95 percent confidence interval:
 0.1646361 0.2838612
sample estimates:
probability of success 
                  0.22 


W tym przypadku wszystkie p-wartości są mniejsze od $0.05$. Odpowiednio dla `prop.test` bez korekcji, z korekcją oraz dla `binom.test` wynoszą one $4.198\cdot 10^{-15},$ $2.383\cdot 10^{-15}$ oraz $6.838\cdot 10^{-16}.$ Odrzucamy więc hipotezę zerową.

### c.
W podpunkcie c rozpatrujemy lek Ibuprom, więc niech $p$ będzie prawdopodobieństwem, że losowo wybrana osoba z badanej populacji w przypadku bólu zażywa Ibuprom.

Z zadania odczytujemy hipotezę zerową jako $$H_0: p \geq \frac{1}{5}$$ oraz hipotezę alternatywną jako $$H_a: p < \frac{1}{5}.$$ 

In [22]:
prop.test(35, 200, p = 1/5, alternative = "l")


	1-sample proportions test with continuity correction

data:  35 out of 200, null probability 1/5
X-squared = 0.63281, df = 1, p-value = 0.2132
alternative hypothesis: true p is less than 0.2
95 percent confidence interval:
 0.0000000 0.2261568
sample estimates:
    p 
0.175 


In [23]:
prop.test(35, 200, p = 1/5, alternative = "l", correct = FALSE)


	1-sample proportions test without continuity correction

data:  35 out of 200, null probability 1/5
X-squared = 0.78125, df = 1, p-value = 0.1884
alternative hypothesis: true p is less than 0.2
95 percent confidence interval:
 0.0000000 0.2234492
sample estimates:
    p 
0.175 


In [24]:
binom.test(35, 200, p = 1/5, alternative = "l")


	Exact binomial test

data:  35 and 200
number of successes = 35, number of trials = 200, p-value = 0.2151
alternative hypothesis: true probability of success is less than 0.2
95 percent confidence interval:
 0.0000000 0.2252414
sample estimates:
probability of success 
                 0.175 


W przypadku p-wartości są większe od $0.05$, (w kolejności: $0.2132$, $0.1884$, $0.2151$). Nie ma więc podstaw do odrzucania hipotezy zerowej.

### d.
W podpunkcie d powtarzamy testy z podpunktów a-c, jednak rozpatrujemy tylko badaną populację do lat 35.

ERROR: Error in library("rlang"): there is no package called 'rlang'


In [36]:
install.packages("data.table")
install.packages("dplyr")
install.packages("formattable")
install.packages("tidyr")

#Load the libraries

library(data.table)
library(dplyr)
library(formattable)
library(tidyr)

#Set a few color variables to make our table more visually appealing

customGreen0 = "#DeF7E9"
customGreen = "#71CA97"
customRed = "#ff7f7f"

Warning message:
"package 'dplyr' was built under R version 4.0.5"


ERROR: Error: package or namespace load failed for 'dplyr' in loadNamespace(i, c(lib.loc, .libPaths()), versionCheck = vI[[i]]):
 namespace 'rlang' 0.4.10 is already loaded, but >= 1.0.0 is required


data
  3  13  18  22  32  38  44  48  54  60  61  70  96 104 113 125 
  1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1 

# Zadanie 3
korzystamy z tablicy 1 
test fishera
mamy sie ograniczyc (ogr tablice)
zapisac jak bd wygladala tabela
rozrozniamy tylko 2 leki
panadol i inny
rozrozniacveyo 2 gru[y  wiekowe

h0, h alt.
fragment z wywolaniem

wynik >> p wart.; wniosek :))

czy na podt... ??? => powolac sie na roznowaznosc hipotez => na podst wynikow tamtego testu mozna wycignac analogiczny wniosek

In [4]:
#?prop.test

# nie ten , tylko test fishera
fisher.test(c(18, 3), c(72, 77))

a <- matrix(c(18, 72, 3, 77), nrow=2, ncol=2)
fisher.test(a)



	Fisher's Exact Test for Count Data

data:  c(18, 3) and c(72, 77)
p-value = 1
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
  0.00000 39.00055
sample estimates:
odds ratio 
         0 


[,1] [,2]
[1,] 18    3  
[2,] 72   77


	Fisher's Exact Test for Count Data

data:  a
p-value = 0.001789
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
  1.748857 35.082794
sample estimates:
odds ratio 
  6.355894 


# Zadanie 4

test chisq dla daych

tylko wywolanie

h0 halt pvalue, wniosek :)

In [ ]:
chisq.test()

In [1]:
?chisq.test

# Zadanie 5
~Kod funkcji w raporcie, nalezy zapisac jak wyglada h0 i ha, ktora weryfikujemy. Nalezy wywloac funkcje dla danych, napisac ile wynosi p-wartosc oraz wniosek. p-wartosc powinna wyjsc 0.21~

Hipoteza zerowa to $$H_0: p \in \mathcal{P}_0,$$ a hipoteza alternatywna to $$H_1: p \in \mathcal{P}_1 = \mathcal{P} \backslash \mathcal{P}_0,$$ gdzie $P_o \subset \mathcal{P} = \{ p: p_i \geq 0, i = 1, \dots, k, \sum_{i=1}^kp_i=1\}$

In [56]:
test_vec <- c(20, 24, 80, 82, 22, 38, 104, 125, 13, 28, 81, 113, 7, 18, 54, 92)

test <- matrix(test_vec, nrow=4)
test

[,1] [,2] [,3] [,4]
[1,] 20    22   13   7  
[2,] 24    38   28  18  
[3,] 80   104   81  54  
[4,] 82   125  113  92

In [58]:
p_val <- function(table) {
    n <- sum(test)
    
    R <- dim(test)[1]
    C <- dim(test)[2]
    
    Gsq <- 1
    
    for (i in 1:R) {
        for (j in 1:C) { #ni+ * n+j
            Gsq <- Gsq * ((sum(table[i, ]) * sum(table[, j])) / (n*table[i, j]))^table[i, j]
        }
    }
    

    Gsq <- -2*log(Gsq)
    
    p_val <- 1 - pchisq(Gsq, (R-1)*(C-1))
    p_val
}

In [59]:
p_val(test)

[1] 0.2112391

P wartość wynosi $0.21.$ Jest ona większa od $0.05$, więc nie odrzucamy hipotezy zerowej (jeśli przyjmujemy poziom istotności $0.05$). Zatem możemy powiedzieć, że zmienne są niezależne.

# Lista 8, 9

## Zadanie 1

W zadaniu pierwszym zaimplementujemy poznane na wykładzie miary współzmienności. 

Zacznijmy jednak od wprowadzenia danych testowych (z wykładu).


In [10]:
#install.packages("DescTools")
#install.packages("psych")
library("DescTools")
library("psych")

test_data <- matrix(c(35, 22, 15, 0, 18, 0, 22, 15, 40, 3, 0, 0, 15, 10, 5), nrow = 5)
test_data

[,1] [,2] [,3]
[1,] 35    0    0  
[2,] 22   22    0  
[3,] 15   15   15  
[4,]  0   40   10  
[5,] 18    3    5

Teraz możemy zaimplementować miary. Dla każdej z nich, napisaliśmy oddzielną funckję, która jest wywoływana przez funcję `coefficient`.

In [16]:
data <- matrix(c(35, 22, 15, 0, 18, 0, 22, 15, 40, 3, 0, 0, 15, 10, 5), nrow = 5)

#1
goodman <- function(data) {
  n <- sum(data)
  
  R <- dim(data)[1]
  C <- dim(data)[2]
  
  sum1 <- 0
  
  for (i in 1:R) {
    for (j in 1:C) {
      sum1 <- sum1 + data[i, j]^2 / (n * sum(data[i, ]))
    }
  }
  
  sum2 <- 0
  for (j in 1:C) {
    sum2 <- sum2 + (sum(data[ , j])/n)^2
  }
  
  (sum1 - sum2) / (1 - sum2)
}

#2
crammer <- function(data) {
  chi2 <- as.numeric(chisq.test(data, simulate.p.value = TRUE)$statistic)
  C <- dim(data)[2]
  R <- dim(data)[1]
  sqrt(chi2/(sum(data)*min(R-1, C-1)))
  
}

#3
t.czuprow <- function(data) {
  chi2 <- as.numeric(chisq.test(data, simulate.p.value = TRUE)$statistic)
  C <- dim(data)[2]
  R <- dim(data)[1]
  sqrt(chi2/(sum(data)*sqrt((R-1)*(C-1))))
}

#4
phi <- function(data) {
  chi2 <- as.numeric(chisq.test(data, simulate.p.value = TRUE)$statistic)
  sqrt(chi2/sum(data))
}

# 5
pearson <- function(data) {
  chi2 <- as.numeric(chisq.test(data, simulate.p.value = TRUE)$statistic)
  sqrt(chi2/(chi2 + sum(data)))
  
}

coefficient <- function(data, coef) {
  if (coef == 1) {
    goodman(data) 
  } else if (coef == 2) {
    crammer(data)
  } else if (coef == 3) {
    t.czuprow(data)
  } else if (coef == 4) {
    phi(data)
  } else if (coef == 5) {
    pearson(data)
  }
}

Sprawdźmy zatem, czy nasze funkcje dziają poprawnie. Porównamy ich wyniki z wynikami funckji w pakiecie R.

In [17]:
coefficient(data, 1)
GoodmanKruskalTau(data, direction = "column")

[1] 0.3477173

[1] 0.3477173

In [24]:
coefficient(data, 2)
CramerV(data)

[1] 0.5361155

[1] 0.5361155

In [20]:
coefficient(data, 3)
TschuprowT(data)

[1] 0.4508176

[1] 0.4508176

In [21]:
coefficient(data, 4)
phi(data)

[1] 0.7581819

[1] 0.7581819

In [22]:
coefficient(data, 5)
ContCoef(data)

[1] 0.6041645

[1] 0.6041645

Jak widać, w każdym przypadku nasza funkcja zwraca dokładnie taką samą wartość.

# Zadanie 2

W zadaniu drugim oraz trzecim przeprowadzamy analizę korespondencji. Zanim jednak do niej przejdziemy, wykonajmy test ..., który zweryfikuje hipotezę o niezależności.


wychodzimy od testu kt bedzie weryfikowal hipoteze o niezaleznosci
wykonac test fishera albo chisq 
przejsc dalej do analizy kor.
przedstawic kod -- funkcje
przedstawi e c wszystkie wyniki
macierz r c dc macierze dekompozycji
wspolczynniki wykres :)

2-3 zd komentarza

In [23]:
# to chyba gdzieś było...

In [1]:
data <- matrix(c(35, 22, 15, 0, 18, 0, 22, 15, 40, 3, 0, 0, 15, 10, 5), nrow = 5)
chisq.test(data)
data

Warning message in chisq.test(data):
"Chi-squared approximation may be incorrect"



	Pearson's Chi-squared test

data:  data
X-squared = 114.97, df = 8, p-value < 2.2e-16


[,1] [,2] [,3]
[1,] 35    0    0  
[2,] 22   22    0  
[3,] 15   15   15  
[4,]  0   40   10  
[5,] 18    3    5

In [3]:
data <- matrix(c(32, 22, 13, 3, 44, 38, 48, 18, 60, 104, 61, 54, 70, 125, 113, 96), nrow = 4)
data

[,1] [,2] [,3] [,4]
[1,] 32   44    60   70 
[2,] 22   38   104  125 
[3,] 13   48    61  113 
[4,]  3   18    54   96

# Zadanie 3

In [ ]:
# test

In [ ]:
# zadanie